polygeist/tools/cgeist/Lib/CGDecl.cc

```diff
@@ -0,0 +1,173 @@
+#include "utils.h"
+
+#include "mlir/Dialect/SYCL/IR/SYCLOps.h"
+#include "mlir/Dialect/SYCL/IR/SYCLOpsDialect.h"
+#include "mlir/Target/LLVMIR/TypeFromLLVM.h"
+
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/WithColor.h"
+
+#define DEBUG_TYPE "CGDecl"
+
+using namespace clang;
+using namespace mlir;
+
+ValueCategory MLIRScanner::VisitVarDecl(clang::VarDecl *Decl) {
+  Decl = Decl->getCanonicalDecl();
+  mlir::Type SubType = Glob.getTypes().getMLIRType(Decl->getType());
+  const unsigned MemType = Decl->hasAttr<clang::CUDASharedAttr>() ? 5 : 0;
+  bool LLVMABI = false, IsArray = false;
+
+  if (Glob.getTypes()
+          .getMLIRType(Glob.getCGM().getContext().getLValueReferenceType(
+              Decl->getType()))
+          .isa<LLVM::LLVMPointerType>())
+    LLVMABI = true;
+  else
+    Glob.getTypes().getMLIRType(Decl->getType(), &IsArray);
+
+  if (!LLVMABI && IsArray)
+    SubType = Glob.getTypes().getMLIRType(
+        Glob.getCGM().getContext().getLValueReferenceType(Decl->getType()));
+
+  LLVM::TypeFromLLVMIRTranslator TypeTranslator(*Module->getContext());
+  ValueCategory InitExpr = nullptr;
+
+  if (Expr *Init = Decl->getInit()) {
+    bool IsVectorType =
+        Glob.getTypes().getMLIRType(Init->getType()).isa<mlir::VectorType>();
+    if ((!isa<clang::InitListExpr>(Init) || IsVectorType) &&
+        !isa<clang::CXXConstructExpr>(Init)) {
+      auto Res = Visit(Init);
+      if (!Res.val) {
+        Decl->dump();
+        assert(Res.val);
+      }
+
+      bool IsReference = Init->isLValue() || Init->isXValue();
+      if (IsReference) {
+        assert(Res.isReference);
+        Builder.create<polygeist::TrivialUseOp>(Loc, Res.val);
+        return Params[Decl] = Res;
+      }
+
+      if (IsArray) {
+        assert(Res.isReference);
+        InitExpr = Res;
+      } else {
+        InitExpr = ValueCategory(Res.getValue(Builder), /*isRef*/ false);
+        if (!InitExpr.val) {
+          Init->dump();
+          assert(false);
+        }
+        SubType = InitExpr.val.getType();
+      }
+    }
+  } else if (auto *Ava = Decl->getAttr<clang::AlignValueAttr>()) {
+    if (auto *Algn = dyn_cast<clang::ConstantExpr>(Ava->getAlignment())) {
+      for (auto *A : Algn->children()) {
+        if (auto *IL = dyn_cast<clang::IntegerLiteral>(A)) {
+          if (IL->getValue() == 8192) {
+            llvm::Type *T =
+                Glob.getCGM().getTypes().ConvertType(Decl->getType());
+            SubType = TypeTranslator.translateType(T);
+            LLVMABI = true;
+            break;
+          }
+        }
+      }
+    }
+  } else if (auto *Ava = Decl->getAttr<clang::InitPriorityAttr>()) {
+    if (Ava->getPriority() == 8192) {
+      llvm::Type *T = Glob.getCGM().getTypes().ConvertType(Decl->getType());
+      SubType = TypeTranslator.translateType(T);
+      LLVMABI = true;
+    }
+  }
+
+  Block *Block = nullptr;
+  Block::iterator Iter;
+  Value Op = createAllocOp(SubType, Decl, MemType, IsArray, LLVMABI);
```

Comment:
originally allocop is only created for the else case, now it is always created, are we going to have more allocs?

Reply:
True, but in the then case `Op` is overwritten. There might be an additional alloc operation in the codegen however it will be dead and so will be removed.

Reply:
I'd rather not include this operation if possible

Reply:
All right will put the else branch back then.

Reply:
done